crates/optimism/trie/src/proof.rs

```diff
@@ -0,0 +1,572 @@
+//! Provides proof operation implementations for [`OpProofsStorage`].
+
+use alloy_primitives::{
+    keccak256,
+    map::{B256Map, HashMap},
+    Address, Bytes, B256, U256,
+};
+use reth_db_api::DatabaseError;
+use reth_execution_errors::{StateProofError, StateRootError, StorageRootError, TrieWitnessError};
+use reth_primitives_traits::Account;
+use reth_trie::{
+    hashed_cursor::{
+        HashedCursor, HashedCursorFactory, HashedPostStateCursorFactory, HashedStorageCursor,
+    },
+    metrics::TrieRootMetrics,
+    proof::{Proof, StorageProof},
+    trie_cursor::{InMemoryTrieCursorFactory, TrieCursor, TrieCursorFactory},
+    updates::TrieUpdates,
+    witness::TrieWitness,
+    AccountProof, BranchNodeCompact, HashedPostState, HashedPostStateSorted, HashedStorage,
+    MultiProof, MultiProofTargets, Nibbles, StateRoot, StorageMultiProof, StorageRoot, TrieInput,
+};
+
+use crate::api::{
+    OpProofsHashedCursor, OpProofsStorage, OpProofsStorageError,
+    OpProofsTrieCursor as OpProofsDBTrieCursor,
+};
+
+/// Manages reading storage or account trie nodes from [`OpProofsDBTrieCursor`].
+#[derive(Debug, Clone)]
+pub struct OpProofsTrieCursor<C: OpProofsDBTrieCursor>(pub C);
+
+impl<C: OpProofsDBTrieCursor> OpProofsTrieCursor<C> {
+    /// Creates a new `OpProofsTrieCursor` instance.
+    pub const fn new(cursor: C) -> Self {
+        Self(cursor)
+    }
+}
+
+impl From<OpProofsStorageError> for DatabaseError {
+    fn from(error: OpProofsStorageError) -> Self {
+        Self::Other(error.to_string())
+    }
+}
+
+impl<C: OpProofsDBTrieCursor + Send + Sync> TrieCursor for OpProofsTrieCursor<C> {
```

Comment:
```suggestion
impl<C> TrieCursor for OpProofsTrieCursor<C> where C: OpProofsDBTrieCursor + Send + Sync {
```
use the where clause syntax when the traits list is more than one trait